File: sdk/php/src/Telemetry.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail;

use GuzzleHttp\Client;

/**
 * 匿名遥测上报
 *
 * 默认开启，可通过环境变量 TEMPMAIL_TELEMETRY_ENABLED=false 或配置关闭。
 * 上报内容仅为操作类型、渠道、成功与否等匿名统计信息，错误消息中的邮箱地址会被脱敏。
 */
final class Telemetry
{
    private const DEFAULT_URL = 'https://sdk-1.openel.top/v1/event';
    private const SDK_VERSION = '0.1.0';

    /** @var Client|null 复用的上报专用 Guzzle 客户端（避免每次上报重建 handler 栈） */
    private static ?Client $client = null;

    /**
     * 获取复用的上报客户端（首次调用时创建并缓存）。
     *
     * 遥测为高频热路径（每次生成邮箱/收信均触发），复用同一客户端可省去
     * 重复构建 handler 栈与中间件的开销，并允许到上报端点的连接 keep-alive 复用。
     */
    private static function client(): Client
    {
        if (self::$client === null) {
            self::$client = new Client(['timeout' => 8]);
        }
        return self::$client;
    }

    /**
     * 判断遥测是否开启。
     */
    private static function enabled(): bool
    {
        $v = ConfigStore::get()->telemetryEnabled;
        return $v === null ? true : $v;
    }

    private static function resolveUrl(): string
    {
        $u = trim((string) (ConfigStore::get()->telemetryUrl ?? ''));
        return $u !== '' ? $u : self::DEFAULT_URL;
    }

    /**
     * 脱敏错误消息中的邮箱地址，并截断长度。
     */
    private static function sanitize(string $msg): string
    {
        if ($msg === '') {
            return '';
        }
        $redacted = preg_replace('/[^\s@]{1,64}@[^\s@]{1,255}/', '[redacted]', $msg) ?? $msg;
        return substr($redacted, 0, 400);
    }

    /**
     * 上报一次操作事件（best-effort，异常静默）。
     */
    public static function report(
        string $operation,
        string $channel,
        bool $success,
        int $attemptCount,
        int $channelsTried,
        string $error,
    ): void {
        if (!self::enabled()) {
            return;
        }

        $ev = [
            'operation' => $operation,
            'channel' => $channel,
            'success' => $success,
            'attempt_count' => $attemptCount,
            'ts_ms' => (int) (microtime(true) * 1000),
        ];
        if ($channelsTried > 0) {
            $ev['channels_tried'] = $channelsTried;
        }
        $err = self::sanitize($error);
        if ($err !== '') {
            $ev['error'] = $err;
        }

        $batch = [
            'schema_version' => 2,
            'sdk_language' => 'php',
            'sdk_version' => self::SDK_VERSION,
            'os' => strtolower(PHP_OS),
            'arch' => php_uname('m'),
            'events' => [$ev],
        ];

        try {
            $body = json_encode($batch, JSON_UNESCAPED_UNICODE);
            self::client()->post(self::resolveUrl(), [
                'body' => $body,
                'headers' => [
                    'Content-Type' => 'application/json; charset=utf-8',
                    'User-Agent' => 'tempmail-sdk-php/' . self::SDK_VERSION,
                ],
                'http_errors' => false,
            ]);
        } catch (\Throwable) {
            // 遥测失败不影响主流程
        }
    }
}

File: sdk/csharp/src/Telemetry.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;

namespace XxxXTeam.TempMail;

/// <summary>
/// 匿名用量遥测。默认开启，可通过配置或环境变量关闭。
/// 上报内容仅含操作类型、渠道、成功与否、尝试次数等，邮箱地址会被脱敏。
/// </summary>
public static class Telemetry
{
    private const string DefaultUrl = "https://sdk-1.openel.top/v1/event";
    private const int MaxBatch = 32;
    private const double FlushSec = 2.0;

    private static readonly Regex EmailRe = new(@"[^\s@]{1,64}@[^\s@]{1,255}", RegexOptions.Compiled);
    private static readonly ConcurrentQueue<Dictionary<string, object?>> Queue = new();
    private static readonly object TimerLock = new();
    private static Timer? _flushTimer;
    private static int _periodicStarted;

    private static bool IsOn()
    {
        var v = Config.Get().TelemetryEnabled;
        return v ?? true;
    }

    private static string ResolveUrl()
    {
        var u = (Config.Get().TelemetryUrl ?? "").Trim();
        return string.IsNullOrEmpty(u) ? DefaultUrl : u;
    }

    private static string Sanitize(string? msg)
    {
        if (string.IsNullOrEmpty(msg)) return "";
        var redacted = EmailRe.Replace(msg, "[redacted]");
        return redacted.Length > 400 ? redacted[..400] : redacted;
    }

    private static string SdkVersion() => "0.1.0";

    /// <summary>上报一条遥测事件（非阻塞，失败静默）</summary>
    public static void Report(string operation, string channel, bool success,
        int attemptCount, int channelsTried, string error)
    {
        if (!IsOn()) return;
        EnsurePeriodic();

        var ev = new Dictionary<string, object?>
        {
            ["operation"] = operation,
            ["channel"] = channel,
            ["success"] = success,
            ["attempt_count"] = attemptCount,
            ["ts_ms"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        };
        if (channelsTried > 0) ev["channels_tried"] = channelsTried;
        var err = Sanitize(error);
        if (!string.IsNullOrEmpty(err)) ev["error"] = err;

        Queue.Enqueue(ev);
        if (Queue.Count >= MaxBatch) FlushBatch();
        else ArmFlushTimer();
    }

    private static void ArmFlushTimer()
    {
        lock (TimerLock)
        {
            // 复用单个 Timer，仅重置到期时间，避免每次入队都 Dispose+new Timer 分配
            if (_flushTimer is null)
                _flushTimer = new Timer(_ => FlushBatch(), null,
                    TimeSpan.FromSeconds(FlushSec), Timeout.InfiniteTimeSpan);
            else
                _flushTimer.Change(TimeSpan.FromSeconds(FlushSec), Timeout.InfiniteTimeSpan);
        }
    }

    private static void EnsurePeriodic()
    {
        if (Interlocked.Exchange(ref _periodicStarted, 1) == 1) return;
        _ = Task.Run(async () =>
        {
            while (true)
            {
                await Task.Delay(TimeSpan.FromSeconds(FlushSec)).ConfigureAwait(false);
                FlushBatch();
            }
        });
    }

    private static void FlushBatch()
    {
        if (!IsOn())
        {
            while (Queue.TryDequeue(out _)) { }
            return;
        }

        var events = new List<Dictionary<string, object?>>();
        while (Queue.TryDequeue(out var ev)) events.Add(ev);
        if (events.Count == 0) return;

        var batch = new Dictionary<string, object?>
        {
            ["schema_version"] = 2,
            ["sdk_language"] = "csharp",
            ["sdk_version"] = SdkVersion(),
            ["os"] = OsName(),
            ["arch"] = System.Runtime.InteropServices.RuntimeInformation.OSArchitecture.ToString().ToLowerInvariant(),
            ["events"] = events,
        };

        var body = Json.Serialize(batch);
        var url = ResolveUrl();
        var ver = SdkVersion();
        _ = Task.Run(() =>
        {
            try
            {
                Http.Post(url, body, "application/json; charset=utf-8",
                    new Dictionary<string, string> { ["User-Agent"] = $"tempmail-sdk-csharp/{ver}" }, 8);
            }
            catch { /* 遥测失败静默 */ }
        });
    }

    private static string OsName()
    {
        var r = System.Runtime.InteropServices.RuntimeInformation.OSDescription.ToLowerInvariant();
        if (r.Contains("windows")) return "windows";
        if (r.Contains("linux")) return "linux";
        if (r.Contains("darwin") || r.Contains("mac")) return "darwin";
        return "unknown";
    }
}
